src/self-managed-debug/src/system_catalog_dumper.rs

```diff
@@ -615,57 +615,231 @@ const RELATIONS: &[Relation] = &[
+        MakeTlsConnector,
+    ),
+    anyhow::Error,
+> {
+    let mut pg_config = PgConfig::from_str(connection_string)?;
+    pg_config.connect_timeout(PG_CONNECTION_TIMEOUT);
+    let tls = make_tls(&pg_config)?;
+    let (pg_client, pg_conn) = retry::Retry::default()
+        .max_duration(PG_CONNECTION_TIMEOUT)
+        .retry_async_canceling(|_| {
+            let pg_config = pg_config.clone();
+            let tls = tls.clone();
+            async move { pg_config.connect(tls).await.map_err(|e| anyhow::anyhow!(e)) }
+        })
+        .await?;
+
+    Ok((pg_client, pg_conn, tls))
+}
+
+pub async fn copy_relation_to_csv(
+    transaction: &Transaction<'_>,
+    file_path_name: PathBuf,
+    column_names: &mut Vec<String>,
+    relation_name: &str,
+) -> Result<(), anyhow::Error> {
+    let mut file = tokio::fs::File::create(&file_path_name).await?;
+
+    file.write_all((column_names.join(",") + "\n").as_bytes())
+        .await?;
+
+    // Stream data rows to CSV
+    let copy_query = format!(
+        "COPY (SELECT * FROM {}) TO STDOUT WITH (FORMAT CSV)",
+        relation_name
+    );
+
+    let copy_stream = transaction
+        .copy_out(&copy_query)
+        .await
+        .context(format!("Failed to COPY TO for {}", relation_name))?
+        .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e));
+    let copy_stream = std::pin::pin!(copy_stream);
+    let mut reader = StreamReader::new(copy_stream);
+    tokio::io::copy(&mut reader, &mut file).await?;
+
+    info!("Copied {} to {}", relation_name, file_path_name.display());
+    Ok::<(), anyhow::Error>(())
+}
+
+pub async fn query_relation(
+    transaction: &Transaction<'_>,
+    start_time: DateTime<Utc>,
+    relation: &Relation,
+    cluster_replica: Option<&ClusterReplica>,
+) -> Result<(), anyhow::Error> {
+    let relation_name = relation.name;
+    let relation_category = &relation.category;
+
+    // Some queries (i.e. mz_introspection relations) require the cluster and replica to be set.
+    if let Some(cluster_replica) = &cluster_replica {
+        transaction
+            .execute(
+                &format!("SET LOCAL CLUSTER = '{}'", cluster_replica.cluster_name),
+                &[],
+            )
+            .await
+            .context(format!(
+                "Failed to set cluster to {}",
+                cluster_replica.cluster_name
+            ))?;
+        transaction
+            .execute(
+                &format!(
+                    "SET LOCAL CLUSTER_REPLICA = '{}'",
+                    cluster_replica.replica_name
+                ),
+                &[],
+            )
+            .await
+            .context(format!(
+                "Failed to set cluster replica to {}",
+                cluster_replica.replica_name
+            ))?;
+    }
+
+    // We query the column names to write the header row of the CSV file.
+    // TODO (SangJunBak): Use `WITH (HEADER TRUE)` once database-issues#2846 is implemented.
```

Comment:
I think this comment is stale? Or it should maybe be moved to `copy_relation_to_csv(...)`?

Reply:
Ah good callout on how it should be closer to copy_relation_to_csv.